wasm-chain-integration/src/v1/trie/low_level.rs

```diff
@@ -1207,31 +1221,38 @@ impl<V> MutableTrie<V> {
             let mut stem_iter = node.path.as_ref().iter();
             match follow_stem(&mut key_iter, &mut stem_iter) {
                 FollowStem::Equal => {
-                    // we lock this node and the entire subtree for modifications.
-                    node.locked = node.locked.checked_add(1).ok_or(TooManyIterators)?;
+                    let generation = node.generation;
+                    if !lock_subtree(&mut iterator_roots[generation as usize], key) {
```

Comment:
Instead of this, make `lock_subtree` return `Result<(), TooManyIterators>`. Then use

```rust
lock_subtree(&mut iterator_roots[generation as usize], key)?
```
This is also where it will help to store the iterators inside the generation_roots. We already looked that up, so we don't need to look up the generation each time.